File: day2/typescript/part_1.ts
import * as fs from 'fs';

const isSafeReport = (report: number[]): boolean => {
    const differences = report.slice(1).map((num, i) => num - report[i]);
    const allIncreasing = differences.every(diff => diff >= 1 && diff <= 3);
    const allDecreasing = differences.every(diff => diff <= -1 && diff >= -3);
    return allIncreasing || allDecreasing;
};

const inputFile = "input.txt";

const fileContent = fs.readFileSync(inputFile, 'utf-8');
const lines = fileContent.trim().split('\n');

let safeCount = 0;

lines.forEach(line => {
    const report = line.split(' ').map(Number);
    if (isSafeReport(report)) {
        safeCount++;
    }
});

console.log(safeCount);

File: day2/rust/part_1.rs
use std::fs;

fn is_safe_report(report: &[i32]) -> bool {
    let differences: Vec<i32> = report.windows(2).map(|w| w[1] - w[0]).collect();
    let all_increasing = differences.iter().all(|&diff| (1..=3).contains(&diff));
    let all_decreasing = differences.iter().all(|&diff| (-3..=-1).contains(&diff));
    all_increasing || all_decreasing
}

fn main() {
    let input_file = "input.txt";
    let contents = fs::read_to_string(input_file).unwrap();
    let lines: Vec<&str> = contents.lines().collect();

    let safe_count = lines.iter()
        .filter(|&&line| {
            let report: Vec<i32> = line.split_whitespace().map(|n| n.parse().unwrap()).collect();
            is_safe_report(&report)
        })
        .count();

    println!("{}", safe_count);
}
